src/charm.py

```diff
@@ -0,0 +1,339 @@
+
+    def configure_kube_ovn(self):
+        if not self.is_kubeconfig_available():
+            self.unit.status = WaitingStatus("Waiting for Kubernetes API")
+            return
+
+        self.check_if_pod_restart_will_be_needed()
+
+        self.apply_crds()
+        self.apply_ovn()
+        self.apply_kube_ovn()
+
+        if self.stored.pod_restart_needed:
+            self.restart_pods()
+
+        self.stored.kube_ovn_configured = True
+
+    def get_container_resource(self, resource, container_name):
+        container = [
+            container
+            for container in resource["spec"]["template"]["spec"]["containers"]
+            if container["name"] == container_name
+        ][0]
+        return container
+
+    def get_ovn_node_ips(self):
+        label = self.model.config["control-plane-node-label"]
+
+        nodes = json.loads(self.kubectl("get", "node", "-l", label, "-o", "json"))[
+            "items"
+        ]
+        node_ips = [
+            address["address"]
+            for node in nodes
+            for address in node["status"]["addresses"]
+            if address["type"] == "InternalIP"
+        ]
+        return node_ips
+
+    def get_resource(self, resources, kind, name):
+        resource = [
+            resource
+            for resource in resources
+            if resource["kind"] == kind and resource["metadata"]["name"] == name
+        ][0]
+        return resource
+
+    def is_kubeconfig_available(self):
+        for relation in self.model.relations["cni"]:
+            for unit in relation.units:
+                if relation.data[unit].get("kubeconfig-hash"):
+                    return True
+        return False
+
+    def kubectl(self, *args):
+        cmd = ["kubectl", "--kubeconfig", "/root/.kube/config"] + list(args)
+        return check_output(cmd)
+
+    def load_manifest(self, name):
+        with open("templates/" + name) as f:
+            return list(yaml.safe_load_all(f))
+
+    def on_cni_relation_joined(self, event):
+        self.configure_cni_relation()
+        self.set_active_status()
+
+    def on_cni_relation_changed(self, event):
+        try:
+            self.configure_kube_ovn()
+        except CalledProcessError:
+            # Likely the Kubernetes API is unavailable. Log a stack trace
+            # and retry.
+            log.error(traceback.format_exc())
+            self.unit.status = WaitingStatus("Waiting to retry configuring Kube-OVN")
+            event.defer()
+            return
+
+        self.set_active_status()
+
+    def on_config_changed(self, event):
+        self.configure_cni_relation()
+
+        try:
+            self.configure_kube_ovn()
+        except CalledProcessError:
+            # Likely the Kubernetes API is unavailable. Log a stack trace
+            # and retry.
+            log.error(traceback.format_exc())
+            self.unit.status = WaitingStatus("Waiting to retry configuring Kube-OVN")
+            event.defer()
```

Comment:
small nit, but this could be refactored into a method since its repeated in the relation changed call as well